### PR TITLE
Increment 5

### PR DESCRIPTION
**Pull request by Taboon**

### CODE REVIEW

File: cmd/shortener/config/config.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
+package config
```
Comment (arttet) [inline cmd/shortener/config/config.go]: Please review `Go` layouts

https://github.com/golang-standards/project-layout

File: cmd/shortener/config/config.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+		l.IP, fmt.Sprint(l.Port),
+	}
+	return fmt.Sprint(strings.Join(address, ":"))
+}
+
+// Set связывает переменную типа со значением флага
+// и устанавливает правила парсинга для пользовательского типа.
+func (l *LocalAddress) Set(flagValue string) error {
+	address := strings.Split(flagValue, ":")
+	l.IP = address[0]
```
Comment (arttet) [inline cmd/shortener/config/config.go]: This is an unsafe operation.

File: cmd/shortener/config/config.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+	}
+	l.Port = port
+	return nil
+}
+
+func (c *Config) URL() string {
+	return c.LocalAddress.IP + ":" + strconv.Itoa(c.LocalAddress.Port)
+}
+
+var ConfigGlobal = Config{
```
Comment (arttet) [inline cmd/shortener/config/config.go]: Please don't use global variables

File: cmd/shortener/main.go [hunk 4]
```diff
@@ -1,3 +1,11 @@
 package main
 
-func main() {}
+import (
+	"github.com/Taboon/urlshortner/cmd/shortener/server"
+)
+
+func main() {
+	if err := server.Run(); err != nil {
+		panic(err)
```
Comment (arttet) [inline cmd/shortener/main.go]: Please don't panic here.

File: cmd/shortener/server/flags.go [hunk 5]
```diff
@@ -0,0 +1,28 @@
+package server
+
+import (
+	"flag"
+	"fmt"
+	"github.com/Taboon/urlshortner/cmd/shortener/config"
```
Comment (arttet) [inline cmd/shortener/server/flags.go]: Please sort imports

File: cmd/shortener/server/flags.go [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+	"os"
+)
+
+func parseFlags() {
```
Comment (arttet) [inline cmd/shortener/server/flags.go]: It would be better if you return `error`

File: cmd/shortener/server/flags.go [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+	if envRunAddr := os.Getenv("SERVER_ADDRESS"); envRunAddr != "" {
+		err := config.ConfigGlobal.LocalAddress.Set(envRunAddr)
+		if err != nil {
+			fmt.Println(err)
+		}
+	}
+	if envBasePath := os.Getenv("RUN_ADDR"); envBasePath != "" {
+		config.ConfigGlobal.BaseURL = envBasePath
+	}
+	fmt.Println("Server started on: " + config.ConfigGlobal.URL())
```
Comment (arttet) [inline cmd/shortener/server/flags.go]: Please use another format way.

File: cmd/shortener/server/handlers.go [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+	"net/http"
+	"strings"
+)
+
+func sendURL(w http.ResponseWriter, r *http.Request) {
+
+	path := r.URL.Path
+	path = strings.Trim(path, "/")
+
+	if v, ok := Stor.CheckID(path); ok {
```
Comment (arttet) [inline cmd/shortener/server/handlers.go]: Please use `Go` style approach

```
if err != nil {
  w.WriteHeader(http.StatusBadRequest)

  return
}
```

File: cmd/shortener/server/server.go [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+	"fmt"
+	"github.com/Taboon/urlshortner/cmd/shortener/config"
+	"github.com/Taboon/urlshortner/cmd/shortener/storage"
+	"github.com/go-chi/chi/v5"
+	"math/rand"
+	"net/http"
+	"strings"
+)
+
+var Stor storage.TempStorage
```
Comment (arttet) [inline cmd/shortener/server/server.go]: Please don't use global variables

File: cmd/shortener/server/server.go [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+	r := chi.NewRouter()
+
+	r.Get("/{id}", sendURL)
+	r.Post("/", getURL)
+	return r
+}
+
+func urlValidator(url string) (string, error) {
+	url = strings.TrimSpace(url)
+	if !strings.HasPrefix(url, "http://") && !strings.HasPrefix(url, "https://") {
```
Comment (arttet) [inline cmd/shortener/server/server.go]: These are magic strings.

File: cmd/shortener/storage/repositories.go [hunk 11]
```diff
@@ -0,0 +1,11 @@
+package storage
+
+type Repositories interface {
```
Comment (arttet) [inline cmd/shortener/storage/repositories.go]: Please add comments

File: cmd/shortener/storage/storage.go [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+	ID  string
+}
+
+type SafeMap struct {
+	mu             sync.Mutex
+	mapStor        map[string]URLData
+	reverseMapStor map[URLData]string
+}
+
+var _ Repositories = (*TempStorage)(nil)
```
Comment (arttet) [inline cmd/shortener/storage/storage.go]:
```suggestion
var _ Repositories = (*tempStorage)(nil)
```

File: cmd/shortener/storage/storage.go [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+	}
+	return nil
+}
+
+func (s TempStorage) CheckID(id string) (URLData, bool) {
+	sm.mu.Lock()
+	defer sm.mu.Unlock()
+
+	// Проверяем, что map был инициализирова н
+	if sm.mapStor == nil {
```
Comment (arttet) [inline cmd/shortener/storage/storage.go]: Please avoid duplicate code.

File: cmd/shortener/storage/storage.go [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+	for _, v := range sm.mapStor {
+		if v.URL == url {
+			return v, true
+		}
+	}
+	return URLData{}, false
+}
+
+func (s TempStorage) RemoveURL(data URLData) error {
+	err := sm.Remove(data)
```
Comment (arttet) [inline cmd/shortener/storage/storage.go]:
```suggestion
	return sm.Remove(data)
```

File: cmd/shortener/storage/storage.go [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+	if sm.mapStor == nil {
+		sm.mapStor = make(map[string]URLData)
+	}
+	if sm.reverseMapStor == nil {
+		sm.reverseMapStor = make(map[URLData]string)
+	}
+
+	// Пишем данные в map
+	_, ok := sm.mapStor[url.ID]
+	if ok {
```
Comment (arttet) [inline cmd/shortener/storage/storage.go]: Please check this code

File: cmd/shortener/storage/storage.go [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+
+type URLData struct {
+	URL string
+	ID  string
+}
+
+type SafeMap struct {
+	mu             sync.Mutex
+	mapStor        map[string]URLData
+	reverseMapStor map[URLData]string
```
Comment (arttet) [inline cmd/shortener/storage/storage.go]: Can you simplify it?

**Review by arttet — CHANGES_REQUESTED** [review]
Please add
- https://about.codecov.io/blog/getting-started-with-code-coverage-for-golang/
- https://github.com/golangci/golangci-lint

File: cmd/shortener/config/config.go (hunk 1 above)
Comment (Taboon) [reply cmd/shortener/config/config.go]: moved the config to the root directory, and refactor structure program

File: cmd/shortener/config/config.go (hunk 2 above)
Comment (Taboon) [reply cmd/shortener/config/config.go]: added a value check before accessing them

File: cmd/shortener/config/config.go (hunk 3 above)
Comment (Taboon) [reply cmd/shortener/config/config.go]: I removed the global variable, added the Server structure, and passed the config through it

File: cmd/shortener/main.go (hunk 4 above)
Comment (Taboon) [reply cmd/shortener/main.go]: Replaced with Log.Fatal

File: cmd/shortener/server/flags.go (hunk 5 above)
Comment (Taboon) [reply cmd/shortener/server/flags.go]: Sorted the imports in the entire project

File: cmd/shortener/server/flags.go (hunk 6 above)
Comment (Taboon) [reply cmd/shortener/server/flags.go]: Redid the output of the function

File: cmd/shortener/server/flags.go (hunk 7 above)
Comment (Taboon) [reply cmd/shortener/server/flags.go]: Changed to fmt.Printf

File: cmd/shortener/server/handlers.go (hunk 8 above)
Comment (Taboon) [reply cmd/shortener/server/handlers.go]: Fixed

File: cmd/shortener/server/server.go (hunk 9 above)
Comment (Taboon) [reply cmd/shortener/server/server.go]: Moved the config to the structure Server

File: cmd/shortener/server/server.go (hunk 10 above)
Comment (Taboon) [reply cmd/shortener/server/server.go]: moved to const

File: cmd/shortener/storage/repositories.go (hunk 11 above)
Comment (Taboon) [reply cmd/shortener/storage/repositories.go]: done

File: cmd/shortener/storage/storage.go (hunk 12 above)
Comment (Taboon) [reply cmd/shortener/storage/storage.go]: I didn't understand what was meant. Do I need to make the TempStorage structure non-exportable?

File: cmd/shortener/storage/storage.go (hunk 13 above)
Comment (Taboon) [reply cmd/shortener/storage/storage.go]: I put the duplicated code in a separate function

File: cmd/shortener/storage/storage.go (hunk 14 above)
Comment (Taboon) [reply cmd/shortener/storage/storage.go]: done

File: cmd/shortener/storage/storage.go (hunk 15 above)
Comment (Taboon) [reply cmd/shortener/storage/storage.go]: conducted a refactoring

File: cmd/shortener/storage/storage.go (hunk 16 above)
Comment (Taboon) [reply cmd/shortener/storage/storage.go]: I can, but it's my engineering solution)

File: cmd/shortener/main.go [hunk 17]
```diff
@@ -1,3 +1,20 @@
...
+import (
+	"github.com/Taboon/urlshortner/config"
+	"github.com/Taboon/urlshortner/internal/server"
+	"github.com/Taboon/urlshortner/internal/storage"
+	"log"
+)
+
+func main() {
+	stor := storage.TempStorage{}
+	serv := server.Server{}
```
Comment (arttet) [inline cmd/shortener/main.go]:
```suggestion
	srv := server.Server{}
```

File: config/config.go [hunk 18]
```diff
@@ -0,0 +1,115 @@
+package config
```
Comment (arttet) [inline config/config.go]: SUGGESTION: it should be in `internal`

File: config/config.go [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+
+	if err != nil {
+		return err
+	}
+	l.Port = port
+	return nil
+}
+
+func (c *Config) URL() string {
+	return c.LocalAddress.IP + ":" + strconv.Itoa(c.LocalAddress.Port)
```
Comment (arttet) [inline config/config.go]: SUGGESTION: please use `fmt.Sprintf`

File: internal/server/server.go [hunk 20]
```diff
@@ -0,0 +1,87 @@
+package server
+
+import (
+	"errors"
+	"fmt"
+	storage2 "github.com/Taboon/urlshortner/internal/storage"
```
Comment (arttet) [inline internal/server/server.go]: SUGGESTION: please replace `storage2`

```
Repository
Warehouse
Depot
Container
Vault
Facility
Stockroom
Archive
Closet
Locker
```

File: internal/server/server.go [hunk 21]
```diff
@@ -0,0 +1,87 @@
...
+		return "", errors.New("is not url")
+	}
+
+	return url, nil
+}
+
+func (s *Server) urlSaver(url string) (string, error) {
+	if _, ok := s.Stor.CheckURL(url); ok {
+		return "", errors.New("url already exist")
+	} else {
```
Comment (arttet) [inline internal/server/server.go]:
```suggestion
	}
```

File: internal/server/server.go [hunk 22]
```diff
@@ -0,0 +1,87 @@
...
+		err := s.Stor.AddURL(urlObj)
+		if err != nil {
+			return "", err
+		}
+		return id, nil
+	}
+}
+
+func (s *Server) generateID() string {
+	ok := true
```
Comment (arttet) [inline internal/server/server.go]: CLARIFICATION: Is this clearer?

```
func RandomString(n int) string {
    const letters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
    rand.Seed(time.Now().UnixNano())
    result := make([]byte, n)
    for i := range result {
        result[i] = letters[rand.Intn(len(letters))]
    }
    return string(result)
}
```

File: internal/storage/storage.go [hunk 23]
```diff
@@ -0,0 +1,117 @@
...
+}
+
+type SafeMap struct {
+	mu             sync.Mutex
+	mapStor        map[string]URLData
+	reverseMapStor map[URLData]string
+}
+
+var _ Repositories = (*TempStorage)(nil)
+var sm = SafeMap{}
```
Comment (arttet) [inline internal/storage/storage.go]: CHANGE REQUEST: please don't use global variables

**Review by arttet — CHANGES_REQUESTED** [review]
Please add the following actions in CI:
- https://about.codecov.io/blog/getting-started-with-code-coverage-for-golang/
- https://github.com/golangci/golangci-lint

File: internal/server/server.go (hunk 22 above)
Comment (Taboon) [reply internal/server/server.go]: refactor

File: internal/server/server.go (hunk 20 above)
Comment (Taboon) [reply internal/server/server.go]: remove storage2

File: config/config.go (hunk 19 above)
Comment (Taboon) [reply config/config.go]: done

File: config/config.go (hunk 18 above)
Comment (Taboon) [reply config/config.go]: done

File: cmd/shortener/main.go (hunk 17 above)
Comment (Taboon) [reply cmd/shortener/main.go]: done

File: internal/storage/storage.go (hunk 23 above)
Comment (Taboon) [reply internal/storage/storage.go]: I don't understand. It's not a variable, it's a structure that I need to create an instance of in another package. If I make it non-exportable, how do I add an instance of it to main.go?

**Comment by arttet** [comment]
```
There is an error processing the coverage reports. Common issues are [files paths](https://docs.codecov.com/docs/fixing-paths)
, empty files or expired reports. See error [reference](https://docs.codecov.com/docs/error-reference)
 page for additional troubleshooting to resolve error.
```

https://app.codecov.io/github/Taboon/urlshortner/commit/a955db78a8c1a669dd29d421f77bcfc70a014199

File: internal/storage/repositories.go [hunk 24]
```diff
@@ -0,0 +1,16 @@
+package storage
+
+type Repositories interface {
```
Comment (arttet) [inline internal/storage/repositories.go]:
```suggestion
type Repository interface {
```

File: internal/storage/storage.go (hunk 23 above)
Comment (arttet) [reply internal/storage/storage.go]: 1. `sm` is a variable.
2. You can create a struct and return an interface.

**Review by arttet — CHANGES_REQUESTED** [review]

File: internal/storage/repositories.go (hunk 24 above)
Comment (Taboon) [reply internal/storage/repositories.go]: done

File: internal/storage/storage.go (hunk 23 above)
Comment (Taboon) [reply internal/storage/storage.go]: done

**Review by arttet — APPROVED** [review]

**Pull request merged**